### PR TITLE
docs: Treesitter

### PR DESCRIPTION
**Pull request by tjdevries**
@vigoux Here is part 1 of generating docs. we can chat about it here. You ditched me while I was on stream waiting for you :crying_cat_face: :cry: :sob: :disappointed: :-1: :frowning_face:

### CODE REVIEW

**Comment by vigoux** [comment]
Damn I missed that ! Sorry !

**Comment by vigoux** [comment]
At first glance, this seems pretty cool ! Thanks for taking care of that !

**Comment by tjdevries** [comment]
`./scripts/gen_vimdoc.py --target treesitter` to generate scripts.

**Comment by vigoux** [comment]
Will work on this soon, hopefully get that merged soon-ish too

**Comment by teto** [comment]
bump, would be nice to merge even a suboptimal PR just so that 0.5 gets some doc.

**Comment by vigoux** [comment]
I'll start working on this soon, just to get things roughly matching the actual api.
Bo time now though...

**Comment by vigoux** [comment]
Hey coming back to this, and I have a few questions.

1. How to I add example code in the doc comments ? That'll make porting current doc easier.
2. Can I have some module level documentation ? We'd need that for things like predicates, etc...
3. Can I rename some things ? Like, `new()` functions should be `LanguageTree.new()` for example.

I think that's all for now, thanks @tjdevries for your work !

**Comment by teto** [comment]
For 1. you can look how it's done for get_count (adding <pre> anchors).
For 2., I believe you can use doxygen anchors to do various things as gen_vimdoc uses doxygen to generate an xml file it parses afterwards. Output is in `tmp-lua-doc/`. the `section` part  is not clear to me.
```
Traceback (most recent call last):
  File "scripts/gen_vimdoc.py", line 1091, in <module>
    main(Doxyfile, args)
  File "scripts/gen_vimdoc.py", line 981, in main
    fail(f'no sections for target: {target}')
  File "scripts/gen_vimdoc.py", line 228, in fail
    raise RuntimeError(s)
RuntimeError: no sections for target: lua
```
if you export DEBUG=1 you can see more output.

**Comment by mjlbach** [comment]
I think it's too early, and it doesn't work well enough to use for anything official yet. Plus I think this is generating docs from lua? TJ has that nvim-treesitter-lua plugin for that which predates babelfish/would be a better fit here eventually.

**Comment by vigoux** [comment]
Okay so this should be pretty much ready. Other docs improvements will come later.

**Review by mhinz — APPROVED** [review]

**Comment by mhinz** [comment]
Looks good to me! Since the doc generator is almost only used by core members anyway, let's merge this soon and fix small potential bugs when they happen. Otherwise this PR will linger around for even longer. :]

File: runtime/lua/vim/treesitter.lua [hunk 1]
```diff
@@ -25,12 +25,12 @@ setmetatable(M, {
...
--- @param bufnr The buffer the parser will be tied to
--- @param lang The language of the parser
--- @param opts Options to pass to the language tree
+---
+--- It is not recommended to use this, use vim.treesitter.get_parser() instead.
+---
+--- @param bufnr The buffer the parser will be tied to
+--- @param lang The language of the parser
+--- @param opts Options to pass to the created language tree
 function M._create_parser(bufnr, lang, opts)
```
Comment (chentau) [inline runtime/lua/vim/treesitter.lua:34]: This isn't that big of a deal, but when I was first trying to read through the source code, I found it confusing that `{create, get}_parser` actually returns a languagetree, which may have multiple parsers. Maybe in the future, a better distinction between language trees and treesitter parsers could be added to the docs.

Comment (vigoux) [reply runtime/lua/vim/treesitter.lua:34]: Indeed this is confusing... We will deprecate those functions in a later pr then, and have function like `setup` to simplify things.

**Comment by tjdevries** [comment]
Seems fine to me. Go for it :)

File: runtime/lua/vim/treesitter.lua (hunk 1 above)
Comment (theHamsta) [reply runtime/lua/vim/treesitter.lua:34]: An alternative would be to rename the language tree to multilanguageparser/parsertree. The language trees have a lot of name clashes with the parser trees they produce. `get_parser()` could maybe be `get_root_parser()` `get_top_level_parser()` `get_parser_tree`. The languagetree still implements the parser API it replaced.

**Comment by teto** [comment]
~~o~~ linting issue ~~o~~

**Comment by vigoux** [comment]
Should be all fixed now !

**Comment by tjdevries** [comment]
OK, I click buttons & people get the docs

**Pull request merged**